examples/solvers/chambolle_pock_denoising_tgv.py

```diff
@@ -0,0 +1,151 @@
+"""Total generalized variation denoising using the Chambolle-Pock solver.
+
+Solves the optimization problem
+
+    min_x ||x - d||_2^2 + TGV_2(x)
+
+Where ``d`` is given noisy data TGV_2 is the second order total generalized
+variation of ``x``, defined as
+
+    TGV_2(x) = min_y lam_1 ||Gx - y||_1 + lam_2 ||Ey||_1
+
+where ``G`` is the spatial gradient operating on the scalar-field ``x`` and
+``E`` is the symmetrized gradient operating on the vector-field ``y``.
+Both one-norms ||.||_1 take the one-norm globally and the two-norm locally,
+i.e. ||y||_1 := sum_i sqrt(sum_j y_i(j)^2) where y_i(j) is the j-th value
+of the vector y_i at location i.
+
+The problem is rewritten as
+
+    min_{x, y} ||x - d||_2^2 +  lam_1 ||Gx - y||_1 + lam_2 ||Ey||_1
+
+which can then be solved with the Chambolle-Pock method.
+
+References
+----------
+[1] K. Bredies and M. Holler. A TGV-based framework for variational image
+decompression, zooming and reconstruction. Part II: Numerics. SIAM Journal
+on Imaging Sciences, 8(4):2851-2886, 2015.
+"""
+
+import numpy as np
+import odl
+
+
+# %%
+# --- Set up the forward operator (identity) --- #
+
+# Reconstruction space: discretized functions on the rectangle
+# [-20, 20]^2 with 300 samples per dimension.
+n = 300
+U = odl.uniform_discr(min_pt=[-20, -20], max_pt=[20, 20], shape=[n, n],
+                      dtype='float32')
+
+# Create the forward operator
+A = odl.IdentityOperator(U)
+
+# %%
+# --- Generate artificial data --- #
+
+# Create phantom
+#phantom = odl.phantom.shepp_logan(U, modified=True)
+
+m = 75
+
+e1 = np.linspace(0, 1, n)
+e2 = np.ones([n, 1])
+x = np.outer(e1, e2)
+
+e1 = np.linspace(1, 0, n-2*m)
+e2 = np.ones([n-2*m, 1])
+x[m:-m, m:-m] = np.outer(e1, e2)
+
+phantom = U.element(x)
+
+phantom.show(title='Phantom')
+
+# Create sinogram of forward projected phantom with noise
+data = A(phantom)
+data += odl.phantom.white_noise(A.range) * np.mean(data) * 0.1
+
+data.show(title='Simulated data (Sinogram)')
+
+# %%
+# --- Set up the inverse problem --- #
+
+# Initialize gradient operator
+G = odl.Gradient(U, method='forward', pad_mode='symmetric')
+V = G.range
+
+Dx = odl.discr.diff_ops.PartialDerivative(U, 0, method='backward',
+                                          pad_mode='symmetric')
+Dy = odl.discr.diff_ops.PartialDerivative(U, 1, method='backward',
+                                          pad_mode='symmetric')
+
+#W = odl.ProductSpace(U, 3, weighting=[1, 1, 2])
```

Comment:
I wrote some comments about this. The code with the weighted spaces is what actually should have been done. It is now marked as TODO